Add option for TorchDispatchMode to ignore torch.compile internals

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #161648

If TorchDispatchMode.ignore_compile_internals() is True, then we turn
off the TorchDispatchMode during the compilation process, instead
turning it back on during runtime of the compiled artifact.

Test Plan:
- new test

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames @Lucaskabela

zou3519 committed:
Add option for TorchDispatchMode to ignore torch.compile internals

If TorchDispatchMode.ignore_compile_internals() is True, then we turn
off the TorchDispatchMode during the compilation process, instead
turning it back on during runtime of the compiled artifact.

Test Plan:
- new test

[ghstack-poisoned]

zou3519 committed:
Update on "Add option for TorchDispatchMode to ignore torch.compile internals"


If TorchDispatchMode.ignore_compile_internals() is True, then we turn
off the TorchDispatchMode during the compilation process, instead
turning it back on during runtime of the compiled artifact.

Test Plan:
- new test

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

bdhirsh commented on test/dynamo/test_misc.py:
```diff
@@ -229,6 +229,61 @@ def boolarg(aa, bb, flag):
+            foo(x)
+
+        self.assertEqual(len(_checksums), 2)
+        # If you are getting "3" here, then that means the .abs().sum()
```
hmm i'm a bit confused by the test - even without your change, wouldnt' the `.abs().sum()` never get compiled? we only call that in the case when our TorchDispatchMode intercepts the custom op foo, but it doesn't look like the test ever calls `mylib.foo` in a compiled region.

zou3519 replied:
I'm just going to delete the comment because I don't know what exactly is going on here. The .abs().sum() isn't being compiled. When Dynamo falls back to eager, it turns out that dynamo decides to compile each of the ops individually (the mul, sin, cos). This means that it had an opportunity to turn itself back on, but I don't know why or how.

zou3519 replied:
This was a bug in my implementation (great catch!)

bdhirsh commented on torch/_dynamo/convert_frame.py:
```diff
@@ -1630,6 +1631,21 @@ def __call__(
+            continue
+        if mode.is_infra_mode():
+            continue
+        return True
```
hmm do you mind explaining what the behavior is when you have multiple modes active, and one has `ignore_compile_internals` set but the other does not?

bdhirsh reviewed (commented):
left some questions, stamp to unblock

zou3519 committed:
Update on "Add option for TorchDispatchMode to ignore torch.compile internals"


If TorchDispatchMode.ignore_compile_internals() is True, then we turn
off the TorchDispatchMode during the compilation process, instead
turning it back on during runtime of the compiled artifact.

Test Plan:
- new test

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

zou3519 committed:
Update on "Add option for TorchDispatchMode to ignore torch.compile internals"


If TorchDispatchMode.ignore_compile_internals() is True, then we turn
off the TorchDispatchMode during the compilation process, instead
turning it back on during runtime of the compiled artifact.

Test Plan:
- new test

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

zou3519 commented:
@pytorchbot merge
